keystone/server/flask/core.py

```diff
@@ -79,6 +79,14 @@
                       'config_file': '/etc/keystone/watcher.yaml',
                       'include_initiator_user_id_in_metric': 'true',
                       'include_target_project_id_in_metric': 'false'}),
+    # CCloud: add rate_limit middleware
```

Comment:
Does the rate_limit middleware inspect a token or auth context somehow? If not, i think that this should be added a little earlier, before `_Middleware(namespace='keystone.server_middleware',` on line 72. The reason is that `build_auth_context` referenced in that middleware will already fetch a lot from the database.

Reply:
1. rate-limit middleware relies on watcher middleware which runs on CADF, which uses both token and auth context.
2. For example, whitelisting works based on :

whitelisting > scope_name_key > domainName/projectName/userName > env_scope = environ.get(‘WATCHER.INITIATOR_PROJECT_ID’) > initiator_project_id = self.get_safe_from_environ(environ, ‘HTTP_X_PROJECT_ID’)
https://github.com/sapcc/openstack-watcher-middleware/blob/master/watcher/watcher.py#L138-L139

3. Hence,  It’s using the HTTP_X_ headers and for authentication requests also from the payload.

Reply:
For 3) The scope from authentication request ist implemented here: https://github.com/sapcc/openstack-watcher-middleware/blob/master/watcher/watcher.py#L383-L405